discovery: Add system-probe endpoint for service list

vitkyrka opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
-->
### What does this PR do?

Add a new endpoint in system-probe to get the service list with the service names. This will be used from the agent check in a future PR, since the agent check doesn't have permission to get this information.

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

https://datadoghq.atlassian.net/browse/USMON-1127
https://datadoghq.atlassian.net/browse/USMON-1126
https://datadoghq.atlassian.net/browse/USMON-1128

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

guyarb commented on pkg/process/net/common_linux.go:
```diff
@@ -15,15 +15,16 @@ import (
+	statsURL              = "http://unix/debug/stats"
+	pprofURL              = "http://unix/debug/pprof"
+	languageDetectionURL  = "http://unix/" + string(sysconfig.LanguageDetectionModule) + "/detect"
+	discoveryOpenPortsURL = "http://unix/" + string(sysconfig.DiscoveryModule) + "/open_ports"
```
Let's add USM as codeowners for the file

hmahmood replied:
Is this endpoint already in system-probe?

guyarb replied:
it is added in this PR

guyarb commented on pkg/process/net/common.go:
```diff
@@ -379,6 +380,30 @@ func (r *RemoteSysProbeUtil) GetPprof(path string) ([]byte, error) {
+		return nil, fmt.Errorf("got non-success status code: path %s, url: %s, status_code: %d", r.path, discoveryOpenPortsURL, resp.StatusCode)
+	}
+
+	res := &discoverymodel.OpenPortsResponse{}
```
maybe out of scope, but I do have a concern about the decoder, it can cause memory explosion

guyarb commented on pkg/collector/corechecks/servicediscovery/model/model.go:
```diff
@@ -0,0 +1,20 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
```
```suggestion
// Copyright 2024-present Datadog, Inc.
```

guyarb replied:
still seeing 2016

guyarb commented on pkg/collector/corechecks/servicediscovery/model/model.go:
```diff
@@ -0,0 +1,20 @@
+// Port represents an open port in the current host.
+type Port struct {
+	PID         int    `json:"pid"`
+	ProcessName string `json:"processName"`
```
```suggestion
	ProcessName string `json:"process_name"`
```

guyarb replied:
process name should be out of scope here
we should map to pid, but not to process name

guyarb commented on pkg/collector/corechecks/servicediscovery/model/model.go:
```diff
@@ -0,0 +1,20 @@
+// Package model contains types for service discovery.
+package model
+
+// Port represents an open port in the current host.
+type Port struct {
```
Let's try to find better naming, as port struct that contains port field - feels like a naming issue

guyarb commented on pkg/collector/corechecks/servicediscovery/servicediscovery.go:
```diff
@@ -278,6 +279,18 @@ func (c *Check) Interval() time.Duration {
 	return refreshInterval
 }
 
+func portsStr(ports []*model.Port) string {
```
let's document

guyarb commented on pkg/collector/corechecks/servicediscovery/module/impl_linux_test.go:
```diff
@@ -0,0 +1,98 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
```
```suggestion
// Copyright 2024-present Datadog, Inc.
```

guyarb commented on pkg/collector/corechecks/servicediscovery/module/impl_linux.go:
```diff
@@ -46,3 +62,30 @@ func (s *discovery) Close() {}
+		return
+	}
+
+	var portsResp []*model.Port
+	for _, p := range ports {
```
```suggestion
	portsResp := make([]*model.Port, len(ports))
	for i, p := range ports {
		portsResp[i] ....
```

guyarb commented on pkg/collector/corechecks/servicediscovery/module/impl_linux.go:
```diff
@@ -6,24 +6,39 @@
 // discovery is an implementation of the Module interface for the discovery module.
-type discovery struct{}
+type discovery struct {
+	portPoller *portlist.Poller
```
document

guyarb commented on pkg/collector/corechecks/servicediscovery/impl_linux.go:
```diff
@@ -87,17 +86,34 @@ func (li *linuxImpl) DiscoverServices() (*discoveredServices, error) {
+		return nil, errWithCode{
+			err:  err,
+			code: errorCodeSystemProbeConn,
+			svc:  nil,
```
redundant 
```suggestion
```

guyarb commented on pkg/collector/corechecks/servicediscovery/impl_linux.go:
```diff
@@ -87,17 +86,34 @@ func (li *linuxImpl) DiscoverServices() (*discoveredServices, error) {
+		}
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	ports, err := sysProbe.GetDiscoveryOpenPorts(ctx)
```
The timeout seems to short.
I think that should be configurable and longer
What is being done in the network check?

guyarb commented on pkg/collector/corechecks/servicediscovery/impl_linux.go:
```diff
@@ -87,17 +86,34 @@ func (li *linuxImpl) DiscoverServices() (*discoveredServices, error) {
-	for _, p := range ports {
-		portsByPID[p.Pid] = append(portsByPID[p.Pid], p)
+
+	portsByPID := map[int][]*model.Port{}
+	for _, p := range ports.Ports {
```
Performance:
```suggestion
	portsByPID := make(map[int][]*model.Port, 0, len(ports.Ports))
	for _, p := range ports.Ports {
```

guyarb commented on pkg/collector/corechecks/servicediscovery/module/impl_linux_test.go:
```diff
@@ -0,0 +1,98 @@
+	workloadmetacomp "github.com/DataDog/datadog-agent/comp/core/workloadmeta/def"
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/servicediscovery/model"
+	"github.com/DataDog/datadog-agent/pkg/util/optional"
+	"github.com/stretchr/testify/require"
```
last import should be moved to the a different import group

guyarb commented on pkg/collector/corechecks/servicediscovery/module/impl_linux_test.go:
```diff
@@ -0,0 +1,98 @@
+	return srv.URL
+}
+
+func startServerAndGetPort(t *testing.T, modURL string) *model.Port {
```
what is the intention of the function?

guyarb commented on pkg/collector/corechecks/servicediscovery/module/impl_linux_test.go:
```diff
@@ -0,0 +1,98 @@
+	})
+	addr := ln.Addr().(*net.TCPAddr)
+
+	req, err := http.NewRequest("GET", modURL+"/discovery/open_ports", nil)
```
use the consts that we have in the code
also use `http.GetMethod`

guyarb reviewed (changes requested):
CRed

hmahmood commented on pkg/collector/corechecks/servicediscovery/module/impl_linux.go:
```diff
@@ -46,3 +70,426 @@ func (s *discovery) Close() {}
+		return
+	}
+
+	if _, err := os.Stat(path.Join(procfs.DefaultMountPoint, pidStr)); os.IsNotExist(err) {
```
The `procfs` path that DD uses can be different from `/proc` and can be set by config. You should use `kernel.HostProc` for this: https://github.com/DataDog/datadog-agent/blob/main/pkg/util/kernel/fs.go#L65

vitkyrka replied:
I've removed this code in refactoring. I think the other stuff should be fine since gopsutil respects `HOST_PROC` and the agent init sets it if necessary?

brycekahle commented on pkg/collector/corechecks/servicediscovery/module/impl_linux.go:
```diff
@@ -46,3 +70,426 @@ func (s *discovery) Close() {}
+	}
+	inode, err := kernel.GetInoForNs(ns)
+	if err != nil {
+		return 0, err
```
`ns` leaks if this error occurs

vitkyrka replied:
Fixed.

brycekahle commented on pkg/collector/corechecks/servicediscovery/module/impl_linux.go:
```diff
@@ -46,3 +70,426 @@ func (s *discovery) Close() {}
+	var sockets []uint64
+	for _, fd := range FDs {
+		if strings.HasPrefix(fd.Path, "socket:[") {
+			sock, err := strconv.Atoi(strings.TrimPrefix(fd.Path[:len(fd.Path)-1], "socket:["))
```
consider `strconv.ParseUint` instead of `strconv.Atoi`

vitkyrka replied:
Done.

vitkyrka commented on pkg/collector/corechecks/servicediscovery/module/impl_linux.go:
```diff
@@ -46,3 +70,426 @@ func (s *discovery) Close() {}
+			if err != nil {
+				continue
+			}
+			if sock < 0 {
```
The < check is gone after implementing Bryce's suggestion. I've not added the == 0 check, not sure it's really needed? The kernel doesn't use inode 0 for these socket inodes since some years ago but looks like an implementation detail: https://patchwork.kernel.org/project/linux-fsdevel/patch/1435245958-4507-1-git-send-email-cmaiolino@redhat.com/

brycekahle commented on pkg/collector/corechecks/servicediscovery/module/impl_linux.go:
```diff
@@ -34,15 +62,359 @@ func (s *discovery) GetStats() map[string]interface{} {
+			if err != nil {
+				continue
+			}
+			sockets = append(sockets, uint64(sock))
```
I think this cast is no longer necessary.

guyarb commented on .github/labeler.yml:
Let's extract it to a different PR

guyarb commented on .gitlab-ci.yml:
Let's extract it to a different PR

guyarb commented on pkg/collector/corechecks/servicediscovery/impl_linux.go:
```diff
@@ -87,19 +85,24 @@ func (li *linuxImpl) DiscoverServices() (*discoveredServices, error) {
+		}
 	}
 
+	log.Debugf("listeners: %v", response.Listeners)
```
How often is DiscoverServices being called?
if once in 1m, then ignore this comment
if more frequent, then we should wrap this call with a check to the log level, as we've discovered in the past that called `log.Debug` when the log level is info or above, still allocates redundant memory.

guyarb commented on pkg/collector/corechecks/servicediscovery/module/impl_linux.go:
```diff
@@ -6,24 +6,52 @@
 // Ensure discovery implements the module.Module interface.
 var _ module.Module = &discovery{}
 
+// cacheData holds process data that should be cached between calls to the
+// endpoint.
+type cacheData struct {
+	env []string
+}
```
Where and How do you ensure the cache cannot bombard the memory of system-probe?

vitkyrka replied:
Now it's only caching the service name. It's only for running, listening processes so it's proportional to that but I'll create a JIRA to check if we need to limit it in any other way as discussed.

guyarb commented on pkg/collector/corechecks/servicediscovery/module/impl_linux.go:
```diff
@@ -34,15 +62,359 @@ func (s *discovery) GetStats() map[string]interface{} {
+	w.WriteHeader(status)
+}
+
+func (s *discovery) handleOpenPorts(w http.ResponseWriter, _ *http.Request) {
```
should we rename the function?

guyarb commented on pkg/collector/corechecks/servicediscovery/module/impl_linux.go:
```diff
@@ -34,15 +62,359 @@ func (s *discovery) GetStats() map[string]interface{} {
+	utils.WriteAsJSON(w, resp)
+}
+
+var allowedEnvironmentVariables = map[string]struct{}{
+	"PATH":                     {},
+	"PWD":                      {},
+	"GUNICORN_CMD_ARGS":        {},
+	"WSGI_APP":                 {},
+	"DD_SERVICE":               {},
+	"DD_TAGS":                  {},
+	"DD_INJECTION_ENABLED":     {},
+	"SPRING_APPLICATION_NAME":  {},
+	"SPRING_CONFIG_LOCATIONS":  {},
+	"SPRING_CONFIG_NAME":       {},
+	"SPRING_PROFILES_ACTIVE":   {},
+	"CORECLR_PROFILER_PATH":    {},
+	"CORECLR_ENABLE_PROFILING": {},
+	"JAVA_TOOL_OPTIONS":        {},
+	"_JAVA_OPTIONS":            {},
+	"JDK_JAVA_OPTIONS":         {},
+	"JAVA_OPTIONS":             {},
+	"CATALINA_OPTS":            {},
+	"JDPA_OPTS":                {},
+	"VIRTUAL_ENV":              {},
+}
```
Should be modified to reflect only the environment variables we actually need to save for service-name extraction

guyarb commented on pkg/collector/corechecks/servicediscovery/module/impl_linux.go:
```diff
@@ -34,15 +62,359 @@ func (s *discovery) GetStats() map[string]interface{} {
+	// sockets have the following pattern "socket:[inode]"
+	var sockets []uint64
+	for _, fd := range FDs {
+		prefix := "socket:["
```
should be declared as a global const to avoid redundant allocations

guyarb commented on pkg/collector/corechecks/servicediscovery/module/impl_linux.go:
```diff
@@ -34,15 +62,359 @@ func (s *discovery) GetStats() map[string]interface{} {
+		}
+		for _, addr := range addrs {
+			if ipnet, ok := addr.(*net.IPNet); ok && !ipnet.IP.IsLoopback() {
+				if ipnet.IP.To4() != nil {
+					out = append(out, ipnet.IP.String())
+				}
```
What about ipv6?

guyarb commented on pkg/collector/corechecks/servicediscovery/module/impl_linux.go:
```diff
@@ -34,15 +62,359 @@ func (s *discovery) GetStats() map[string]interface{} {
+// pkg/security/security_profile/activity_tree/process_node_snapshot.go.  Get
+// the list of open ports with socket inodes for the provided namespace.
+func getNsInfo(pid int, isRoot bool) (*namespaceInfo, error) {
+	path := filepath.Join(kernel.HostProc(fmt.Sprintf("%d", pid)))
```
Is `Join` actually relevant here?

guyarb commented on pkg/collector/corechecks/servicediscovery/module/impl_linux.go:
```diff
@@ -34,15 +62,359 @@ func (s *discovery) GetStats() map[string]interface{} {
+		socketInodeToPort: make(map[uint64]int),
+	}
+
+	for _, sock := range TCP {
+		if sock.St != tcpListen {
+			continue
+		}
+		nsInfo.socketInodeToPort[sock.Inode] = int(sock.LocalPort)
+	}
+	for _, sock := range TCP6 {
+		if sock.St != tcpListen {
+			continue
+		}
+		nsInfo.socketInodeToPort[sock.Inode] = int(sock.LocalPort)
+	}
+	for _, sock := range UDP {
+		if sock.St != udpListen {
+			continue
+		}
+		nsInfo.socketInodeToPort[sock.Inode] = int(sock.LocalPort)
+	}
+	for _, sock := range UDP6 {
+		if sock.St != udpListen {
+			continue
+		}
+		nsInfo.socketInodeToPort[sock.Inode] = int(sock.LocalPort)
+	}
```
nit: Can that be generalized and reduce redundant duplication?

guyarb commented on pkg/collector/corechecks/servicediscovery/module/impl_linux.go:
```diff
@@ -34,15 +62,359 @@ func (s *discovery) GetStats() map[string]interface{} {
+	}
+}
+
+func (s *discovery) cleanCache(alivePids map[int32]struct{}) {
+	for pid := range s.cache {
+		if _, alive := alivePids[pid]; alive {
+			continue
+		}
+
+		delete(s.cache, pid)
+	}
```
AFAIK, deleting from a map does not free up the memory

vitkyrka replied:
It should free the memory for the strings, right? I added a comment.

guyarb commented on pkg/collector/corechecks/servicediscovery/module/impl_linux.go:
general comment: Let's add documentation here for the different functions to make it clear for the others

guyarb commented:
Only a suggestion - I'd split this PR to smaller PRs as it contains many changes, but those are not "atomic" changes.

guyarb commented on pkg/collector/corechecks/servicediscovery/model/model.go:
```diff
@@ -0,0 +1,20 @@
+type Listener struct {
+	PID     int      `json:"pid"`
+	Name    string   `json:"string"`
+	Environ []string `json:"environ"`
```
```suggestion
	Environ []string `json:"environ,omitempty"`
```

guyarb commented on pkg/collector/corechecks/servicediscovery/model/model.go:
```diff
@@ -0,0 +1,20 @@
+	PID     int      `json:"pid"`
+	Name    string   `json:"string"`
+	Environ []string `json:"environ"`
+	CWD     string   `json:"cwd"`
```
An improvement - 
Let's treat `/` as the default value, and if it is, then we should not explicitly set it, and by that, sparing more redundant characters

guyarb replied:
Why do we need CWD?

guyarb commented on pkg/collector/corechecks/servicediscovery/module/impl_linux.go:
```diff
@@ -34,15 +62,359 @@ func (s *discovery) GetStats() map[string]interface{} {
+		return nil
+	}
+
+	name, err := proc.Name()
```
this is wrong, it returns the process name, instead of the service name

vitkyrka replied:
The service name was being calculated in the agent check but now it's moved here.

guyarb commented on pkg/collector/corechecks/servicediscovery/model/model.go:
```diff
@@ -0,0 +1,20 @@
+// Listener represents a listening process.
+type Listener struct {
+	PID     int      `json:"pid"`
+	Name    string   `json:"string"`
```
```suggestion
	Name    string   `json:"name"`
```

vitkyrka committed:
discovery: Add system-probe endpoint to get service

Add an endpoint which can be used by the agent check to get a list of
services along with their service names.

vitkyrka committed:
discovery: Export ServiceDetector

It will be need to be used from the system-probe component which is in a
different module.

vitkyrka committed:
discovery: Add GetServiceName() to service detector

Add a function to only get the service name without any of the other
metadata, to be used from the initial version of the system-probe
endpoint.

vitkyrka commented:
> Only a suggestion - I'd split this PR to smaller PRs as it contains many changes, but those are not "atomic" changes.

CI changes split to https://github.com/DataDog/datadog-agent/pull/28193
This PR is to add the endpoint to system-probe.
Next PR will use it in agent check.
Deletion of portlist will be done separately.

guyarb commented on pkg/collector/corechecks/servicediscovery/module/impl_linux.go:
```diff
@@ -34,15 +63,261 @@ func (s *discovery) GetStats() map[string]interface{} {
+			sockets = append(sockets, sock)
+		}
+	}
+	if len(sockets) <= 0 {
+		return nil, nil
+	}
+
+	return sockets, nil
```
nit: shouldn't `return socket, nil` give the same outcome?

guyarb commented on pkg/collector/corechecks/servicediscovery/module/impl_linux.go:
```diff
@@ -34,15 +63,261 @@ func (s *discovery) GetStats() map[string]interface{} {
+	if err != nil {
+		return nil
+	}
+	if sockets == nil {
```
maybe `len(sockets) == 0` will be better

guyarb commented on pkg/collector/corechecks/servicediscovery/module/impl_linux.go:
```diff
@@ -34,15 +63,261 @@ func (s *discovery) GetStats() map[string]interface{} {
+		return "", nil
+	}
+
+	env, err := proc.Environ()
```
(can be done in a followup PR)

vitkyrka committed:
discovery: Simplify slice checks

vitkyrka commented:
/merge
